sa/sa.go

```diff
@@ -210,41 +210,38 @@ func (ssa *SQLStorageAuthority) GetRegistrationByKey(ctx context.Context, key *j
 
 // GetAuthorization obtains an Authorization by ID
 func (ssa *SQLStorageAuthority) GetAuthorization(ctx context.Context, id string) (core.Authorization, error) {
-	authz := core.Authorization{}
-	tx, err := ssa.dbMap.Begin()
-	if err != nil {
-		return authz, err
-	}
-	txWithCtx := tx.WithContext(ctx)
-
-	pa, err := selectPendingAuthz(txWithCtx, "WHERE id = ?", id)
-	if err != nil && err != sql.ErrNoRows {
-		return authz, Rollback(tx, err)
-	}
-	if err == sql.ErrNoRows {
-		var fa authzModel
-		err := txWithCtx.SelectOne(&fa, fmt.Sprintf("SELECT %s FROM authz WHERE id = ?", authzFields), id)
+	authz, overallError := withTransaction(ctx, ssa.dbMap, func(txWithCtx transaction) (interface{}, error) {
+		pa, err := selectPendingAuthz(txWithCtx, "WHERE id = ?", id)
 		if err != nil && err != sql.ErrNoRows {
-			return authz, Rollback(tx, err)
-		} else if err == sql.ErrNoRows {
-			// If there was no result in either the pending authz table or the authz
-			// table then return a `berrors.NotFound` instance (or a rollback error if
-			// the transaction rollback fails)
-			return authz, Rollback(
-				tx,
-				berrors.NotFoundError("no authorization found with id %q", id))
-		}
-		authz = fa.Authorization
-	} else {
-		authz = pa.Authorization
-	}
+			return nil, err
+		}
+		var authz core.Authorization
+		if err == sql.ErrNoRows {
+			var fa authzModel
+			err := txWithCtx.SelectOne(&fa, fmt.Sprintf("SELECT %s FROM authz WHERE id = ?", authzFields), id)
+			if err != nil && err != sql.ErrNoRows {
+				return nil, err
+			} else if err == sql.ErrNoRows {
+				// If there was no result in either the pending authz table or the authz
+				// table then return a `berrors.NotFound` instance
+				return nil, berrors.NotFoundError("no authorization found with id %q", id)
+			}
+			authz = fa.Authorization
+		} else {
+			authz = pa.Authorization
+		}
 
-	authz.Challenges, err = ssa.getChallenges(txWithCtx, authz.ID)
-	if err != nil {
-		return authz, Rollback(tx, err)
-	}
+		authz.Challenges, err = ssa.getChallenges(txWithCtx, authz.ID)
+		if err != nil {
+			return nil, err
+		}
 
-	return authz, tx.Commit()
+		return authz, nil
+	})
+	if overallError != nil {
+		return core.Authorization{}, overallError
+	}
+	return authz.(core.Authorization), nil
```

Comment:
Like we chatted about in the comment thread this (and the other similar locations) should check the cast result.